### ISSUE TITLE
Some questions about contrastive learning loss

### ISSUE THREAD

**Issue by Renp1ngs** [Haochen-Wang409/U2PL#16]
https://github.com/Haochen-Wang409/U2PL/blob/35e3c2279922a05872f85457c2dc5f97b6e9ec6a/train_semi.py#L516

May I ask why the loss here needs to be divided by the number of threads.

**Comment by Haochen-Wang409** [comment]
Hi, this is because ```dist.all_reduce``` returns the sum of losses on different GPUs.
Please refer to [this link](https://pytorch.org/docs/stable/distributed.html?highlight=all_reduce#torch.distributed.all_reduce) to see the docs.

**Comment by Renp1ngs** [comment]
thank you for your reply. So why don't you need to do this only for sup_loss and unsup_loss?

> Hi, this is because `dist.all_reduce` returns the sum of losses on different GPUs. Please refer to [this link](https://pytorch.org/docs/stable/distributed.html?highlight=all_reduce#torch.distributed.all_reduce) to see the docs.

**Comment by Renp1ngs** [comment]
thank you for your reply. So why don't you need to do this for sup_loss and unsup_loss?

**Comment by Haochen-Wang409** [comment]
This is because when we compute contrastive loss on different GPU, sometimes there are not any qualified query on some particular GPUs, contrastive losses are returns very fast on these GPUs. 
However, on other GPUs, it will take a little time in computing contrastive loss, resulting in a bug in DDP, and we will not get any output since then...

Therefore, we use ```dist.all_reduce``` to enforce that every GPU will take a little time in backpropagating the contrastive loss.

**Comment by Renp1ngs** [comment]
Satisfactory answer. Although I don't fully understand hhh. Thank you very much.

**Comment by Haochen-Wang409** [comment]
Maybe you could try training without ```dist.all_reduce``` and print ```loss_contra``` on each GPU. 
And then you may understand what I have said hhhh.
